storage/operation/headers.go

```diff
@@ -16,16 +18,20 @@ func RetrieveHeader(r storage.Reader, blockID flow.Identifier, header *flow.Head
 	return RetrieveByKey(r, MakePrefix(codeHeader, blockID), header)
 }
 
-// IndexBlockHeight indexes the height of a block. It should only be called on
-// finalized blocks.
-// TODO(leo): add synchronization
-func IndexBlockHeight(rw storage.ReaderBatchWriter, height uint64, blockID flow.Identifier) error {
+// IndexBlockHeight indexes the height of a block. It must only be called on finalized blocks.
+// This function guarantees that the index is only inserted once for each height.
+// The caller must acquire the [storage.LockFinalizeBlock] lock.
+// Returns [storage.ErrAlreadyExists] if an ID has already been finalized for this height.
+func IndexBlockHeight(lctx lockctx.Proof, rw storage.ReaderBatchWriter, height uint64, blockID flow.Identifier) error {
+	if !lctx.HoldsLock(storage.LockFinalizeBlock) {
+		return fmt.Errorf("missing required lock: %s", storage.LockFinalizeBlock)
```

Comment:
Maybe the HoldsLock method could include the lock name in the error message, so that we just need to wrap data that didn't get passed to the HoldsLock method:
```suggestion
		return fmt.Errorf("missing required lock for block %v (height: %v): %w", blockID, height, err)
```

Reply:
👍  I like this idea. Will save a lot of boilerplate.